### PR TITLE
chore: add ADR for header handling

### PR DESCRIPTION
**Pull request by FrankEssenberger**
Relates to: SAP/cloud-sdk-backlog#361
Relates to: SAP/cloud-sdk-backlog#404

ADR for how to handle the headers in a consistent way.

### CODE REVIEW

File: knowledge-base/adr/0029-customer-header-handling.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+1. Custom headers set by the `addCustomHeaders()` on the [odata-request-config.ts](../../packages/core/src/odata-common/request/odata-delete-request-config.ts) or [openapi-request-builder.ts](../../packages/core/src/openapi/openapi-request-builder.ts)
+2. Destination related headers. These come in two flavors:
+   1. Auth tokens for the target system and proxy headers from the destination and connectivity service [buildAuthorizationHeaders](../../packages/connectivity/src/scp-cf/authorization-header.ts)
+   2. Based on destination properties set by the user (see [getAdditionalHeadersAndQueryParameters](../../packages/connectivity/src/scp-cf/destination/destination.ts).
+3. SDK related headers like `eTag` (only OData) found in the `headers()` method in the [odata-request.ts](../../packages/core/src/odata-common/request/odata-request.ts).
+
+The problem is also present to a smaller extent for query parameters, which can come from:
+
+1. A [custom setting](../../packages/core/src/odata-common/request/odata-request.ts) on request
+2. [Sestination properties](../../packages/connectivity/src/scp-cf/destination/destination.ts)
```
Comment (jjtang1985) [inline knowledge-base/adr/0029-customer-header-handling.md:22]:
```suggestion
2. [Destination properties](../../packages/connectivity/src/scp-cf/destination/destination.ts)
```

File: knowledge-base/adr/0029-customer-header-handling.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+3. Queries like get byt key i.e the SDK.
```
Comment (jjtang1985) [inline knowledge-base/adr/0029-customer-header-handling.md:23]:
```suggestion
3. Queries like get by key i.e the SDK.
```

File: knowledge-base/adr/0029-customer-header-handling.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+ */
+export function executeHttpRequestInternal(config:ConfigInternal){
+    ...
+}
+
+/**
+ * This method is for direct customer use -> provided headers are custom and have high prio.
+ */
+export function executeHttpRequest(config:ConfigPublic){
+    const withObject = headerObjectWithOriginCustom(config) ///
```
Comment (jjtang1985) [inline knowledge-base/adr/0029-customer-header-handling.md:101]:
```suggestion
    const withObject = headerObjectWithOriginCustom(config)
```

File: knowledge-base/adr/0029-customer-header-handling.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+## Breaking Changes
+
+We investigate the risk of breaking changes:
+
+- Changing the type of the `header` property from `string|any` to something more structured i.e. `string|HeaderValueObject` is possible (Option A).
+- Extending the interface with new properties is possible - the custom ones remain where they are (Option B).
+- Removing the optional `customHeader` argument for the methods in [authentication-header.ts](<(../../packages/connectivity/src/scp-cf/authorization-header.ts)>) is not possible.
+- There is a risk of intrinsic behavior changes.
+
+## Decision
```
Comment (jjtang1985) [inline knowledge-base/adr/0029-customer-header-handling.md:115]: The user is working on a live system with modified SDK as a work around.
We should apply this and fix the bug on 2.0 branch.
I added this info to the follow up ticket.

**Review by jjtang1985 — APPROVED** [review]
LGTM.
I added some minor comments.

File: knowledge-base/adr/0029-customer-header-handling.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    origin:'Custom'|'DestinationProperty'|'Destination'|'SDK'
+}
+
+//Parameters are handled analogously and are skipped in the methods to be shorted.
+interface ParameterValueObject {
+    value:string
+    origin:'Custom'|'DestinationProperty'|'SDK'
+}
+
+interface ConfigPublic{
```
Comment (marikaner) [inline knowledge-base/adr/0029-customer-header-handling.md:76]: Lint not working again?
```suggestion
interface ConfigPublic {
```

**Pull request merged**